bwt/bwt.go

```diff
@@ -0,0 +1,227 @@
+// BWT Burrow Wheeler Transform
+// Compresses and Indexes a given sequence so that it can be
+// be used for search, alignment, and text extraction. This is
+// useful for sequences so large that it would be beneficial
+// to reduce its memory footprint while also maintaining a way
+// to analyze and work with the sequence.
+type BWT struct {
+	// firstColumnSkipList is the first column of the BWT. It is
+	// represented as a list of skipEntries because the first column of
+	// the BWT is always lexographically ordered. This saves time and memory.
+	firstColumnSkipList []skipEntry
+	// lastCoulmn last column of the BWT- the actual textual representation
+	// of the BWT.
+	lastCoulmn waveletTree
+	// suffixArray an array that allows us to map a posistion in the first
+	// column to a position in the original sequence. This is needed to be
+	// able to extract text from the BWT.
+	suffixArray []int
+}
+
+// Count represents the number of times the provided pattern
+// shows up in the original sequence.
+func (bwt BWT) Count(pattern string) int {
+	searchRange := bwt.lfSearch(pattern)
+	return searchRange.end - searchRange.start
+}
+
+// Locate returns a list of offsets at which the begging
+// of the provided pattern occurrs in the original
+// sequence.
+func (bwt BWT) Locate(pattern string) []int {
+	searchRange := bwt.lfSearch(pattern)
+	if searchRange.start >= searchRange.end {
+		return nil
+	}
+
+	numOfOffsets := searchRange.end - searchRange.start
+	offsets := make([]int, numOfOffsets)
+	for i := 0; i < numOfOffsets; i++ {
+		offsets[i] = bwt.suffixArray[searchRange.start+i]
+	}
+
+	return offsets
+}
+
+// Extract this allows us to extract parts of the original
+// sequence from the BWT.
+// start is the begging of the range of text to extract inclusive.
+// end is the end of the range of text to extract exclusive.
+// If either start or end are out of bounds, Extract will panic.
+func (bwt BWT) Extract(start, end int) string {
+	if end > bwt.getLenOfOriginalString() {
+		msg := fmt.Sprintf("end [%d] exceeds the max range of the BWT [%d]", end, bwt.getLenOfOriginalString())
+		panic(msg)
+	}
+	if start < 0 {
+		msg := fmt.Sprintf("start [%d] exceeds the min range of the BWT [0]", start)
+		panic(msg)
+	}
+
+	strB := strings.Builder{}
+	for i := start; i < end; i++ {
+		fPos := bwt.getFCharPosFromOriginalSequenceCharPos(i)
+		skip := bwt.lookupSkipByOffset(fPos)
+		strB.WriteByte(skip.char)
+	}
+	return strB.String()
+}
+
+// getFCharPosFromOriginalSequenceCharPos looks up mapping from the original position
+// of the sequence to its corresponding posisiton in the First Column of the BWT
+func (bwt BWT) getFCharPosFromOriginalSequenceCharPos(originalPos int) int {
+	for i := range bwt.suffixArray {
+		if bwt.suffixArray[i] == originalPos {
+			return i
+		}
+	}
+	panic("Unable to find the corresponding orginal positiong for a character in the original sequence in the suffix array. This should not be possible and indicates a malformed BWT.")
```

Comment:
this is a little meta but what are your thoughts on panics and when to use them? I almost always try to return an error and have it be handled upstream for fear of killing a user'ss long run.

Reply:
It depends. I'm usually very hesitant to put them in anywhere as well. However in these cases in the PR, the panics are analogous to having an index out of bounds error for an array. If something wrong is happening like this in the internals of any of these data structures, then something has gone horribly wrong and the program can't recover since the BWT is likely borked. 

That or if whoever is using is upstream is using it incorrectly, they can't recover from that incorrectness. Like in the case of the panic on the out of bounds on Extract. Imagine having to handle errors from index out of bounds.

Reply:
Thought about this some more... Also, sounds like the Go team has some opinions on this: https://go.dev/blog/defer-panic-and-recover

It does seem weird to bubble up errors like this because it is an internal problem, but Poly is also the first open source project I've contributed to- perhaps its best to give the users of Poly the choice of how they want to handle potential errors.

Maybe the example of the json package they mentioned in the blog post would also be appropriate here. We can panic internally and recover at the API boundary. That or we could just bubble errors all the way up for safety if for some reason we ever wanted to reuse the auxiliary components in here so we don't have any hidden panics.

Happy to change it up!

Reply:
Still happy to change it up. With that said, I think the panics in bitvector are appropriate.

Reply:
Started changing it up in f5b459b07749bf1b4f45f7f2d034d23cd97aaf5b. LMK What you think.